gdal/ogr/ogrsf_frmts/ngw/ogrngwlayer.cpp

```diff
@@ -0,0 +1,1116 @@
+{
+    return FeatureToJson(poFeature).Format(CPLJSONObject::Plain);
+}
+
+ /*
+  * OGRNGWLayer()
+  */
+OGRNGWLayer::OGRNGWLayer( OGRNGWDataset *poDSIn,
+    const CPLJSONObject &oResourceJsonObject ) :
+    osResourceId(oResourceJsonObject.GetString("resource/id", "-1")),
+    poDS(poDSIn),
+    bFetchedPermissions(false),
+    nFeatureCount(-1),
+    oNextPos(moFeatures.begin()),
+    nPageStart(0),
+    bNeedSyncData(false),
+    bNeedSyncStructure(false)
+{
+    std::string osName = oResourceJsonObject.GetString("resource/display_name");
+    poFeatureDefn = new OGRFeatureDefn( osName.c_str() );
+    poFeatureDefn->Reference();
+
+    poFeatureDefn->SetGeomType( NGWAPI::NGWGeomTypeToOGRGeomType(
+        oResourceJsonObject.GetString("vector_layer/geometry_type")) );
+
+    OGRSpatialReference *poSRS = new OGRSpatialReference();
+    int nEPSG = oResourceJsonObject.GetInteger("vector_layer/srs/id", 3857); // Default NGW SRS is Web mercator EPSG:3857.
+    if( poSRS->importFromEPSG( nEPSG ) == OGRERR_NONE )
+    {
+        if( poFeatureDefn->GetGeomFieldCount() != 0 )
+        {
+            poFeatureDefn->GetGeomFieldDefn(0)->SetSpatialRef( poSRS );
+        }
+    }
+    else
+    {
+        delete poSRS;
+    }
+
+    CPLJSONArray oFields = oResourceJsonObject.GetArray("feature_layer/fields");
+    FillFields( oFields );
+    FillMetadata( oResourceJsonObject );
+
+    SetDescription( poFeatureDefn->GetName() );
+}
+
+/*
+ * OGRNGWLayer()
+ */
+OGRNGWLayer::OGRNGWLayer( OGRNGWDataset *poDSIn, const std::string &osNameIn,
+    OGRSpatialReference *poSpatialRef, OGRwkbGeometryType eGType,
+    const std::string &osKeyIn, const std::string &osDescIn ) :
+    osResourceId("-1"),
+    poDS(poDSIn),
+    bFetchedPermissions(false),
+    nFeatureCount(0),
+    oNextPos(moFeatures.begin()),
+    nPageStart(0),
+    bNeedSyncData(false),
+    bNeedSyncStructure(false)
+{
+    poFeatureDefn = new OGRFeatureDefn( osNameIn.c_str() );
+    poFeatureDefn->Reference();
+
+    poFeatureDefn->SetGeomType( eGType );
+
+    if( poSpatialRef )
+    {
+        if( poFeatureDefn->GetGeomFieldCount() != 0 )
+        {
+            poFeatureDefn->GetGeomFieldDefn(0)->SetSpatialRef( poSpatialRef );
+        }
+    }
+
+    if( !osDescIn.empty() )
+    {
+        OGRLayer::SetMetadataItem( "description", osDescIn.c_str() );
+    }
+    if( !osKeyIn.empty() )
+    {
+        OGRLayer::SetMetadataItem( "keyname", osKeyIn.c_str() );
+    }
+
+    SetDescription( poFeatureDefn->GetName() );
+}
+
+/*
+ * ~OGRNGWLayer()
+ */
+OGRNGWLayer::~OGRNGWLayer()
+{
+    if( !soChangedIds.empty() )
+    {
+        bNeedSyncData = true;
+    }
+    SyncFeatures();
+    FreeFeaturesCache();
+    if( poFeatureDefn != nullptr )
+    {
+        poFeatureDefn->Release();
+    }
+}
+
+/*
+ * FreeFeaturesCache()
+ */
+void OGRNGWLayer::FreeFeaturesCache()
+{
+    for (std::map<GIntBig, OGRFeature*>::iterator it = moFeatures.begin();
```

Comment:
you could use for( auto& pair: moFeatures )

Reply:
Fix per 4579aeab927d70ddab6142340bf5550c6789b42e